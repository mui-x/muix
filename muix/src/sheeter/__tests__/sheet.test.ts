import {
  atomizeSheet,
  atomizeRuleset,
  mergeSheets,
} from "reactxx-sheeter"
import { initPlatform, Shape, theme } from "reactxx-tests"

describe("SHEET", () => {
  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let sheet
    it("01 null", () => {
      sheet = atomizeSheet(null)
      expect(sheet).toMatchSnapshot()
    })
    it("02 {}", () => {
      sheet = atomizeSheet({})
      expect(sheet).toMatchSnapshot()
    })
    it("03 backgroundColor: 'red'", () => {
      sheet = atomizeSheet<Shape>({root: {backgroundColor: 'red'}})
      expect(sheet).toMatchSnapshot()
    })
    it("04 merge", () => {
      sheet = mergeSheets(null)
      expect(sheet).toMatchSnapshot()
    })
    it("05 merge", () => {
      sheet = mergeSheets([])
      expect(sheet).toMatchSnapshot()
    })
    it("06 merge", () => {
      sheet = mergeSheets([null, {}, undefined])
      expect(sheet).toMatchSnapshot()
    })
    it("07 merge", () => {
      sheet = mergeSheets([
        null, {}, undefined,
        atomizeSheet<Shape>({
          root: {backgroundColor: 'green'},
          label: {color: 'green'},
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("08 merge", () => {
      sheet = mergeSheets([
        atomizeSheet<Shape>({
          root: {backgroundColor: 'red'},
          label: {color: 'red'},
        }),
        atomizeSheet<Shape>({
          root: {backgroundColor: 'green'},
          webOnly: { $web: {color: 'green'}},
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("09 merge", () => {
      sheet = mergeSheets([
        atomizeSheet<Shape>({
          root: {backgroundColor: 'red'},
        }),
        atomizeSheet<Shape>({
          root: { 
            $web: {color: 'green'},
          }
        }),
        atomizeSheet<Shape>({
          root: { 
            $native: {backgroundColor: 'blue'},
          }
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("10 with theme", () => {
      sheet = atomizeSheet<Shape>(theme => ({
        root: {
          backgroundColor: theme.primary.normal.backgroundColor
        },
        label: [
          theme.primary.normal,
          {margin: 10}
        ],
        webOnly: {
          $web: atomizeRuleset<'Text', Shape>(theme.secondary.disabled, null, 'theme.secondary.disabled')
        }
      }), theme)
      expect(sheet).toMatchSnapshot()
    })
    it("11 sheet source mutated", () => {
      const sheetSource = {
        root: {
          backgroundColor: 'red',
          $web: { color: 'green' }
        }
      }
      expect(sheetSource).toMatchSnapshot()
      sheet = atomizeSheet<Shape>(sheetSource)
      expect(sheetSource).toMatchSnapshot()
      expect(sheet).toMatchSnapshot()
      sheet = atomizeSheet<Shape>(sheet)
      expect(sheet).toMatchSnapshot()
    })
    it("12 merge ERROR, sheets must be atomized first", () => {
      sheet = mergeSheets([
        {
          root: {backgroundColor: 'red'},
          label: {color: 'red'},
        },
        atomizeSheet<Shape>({
          root: {backgroundColor: 'green'},
          webOnly: { $web: {color: 'green'}},
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
