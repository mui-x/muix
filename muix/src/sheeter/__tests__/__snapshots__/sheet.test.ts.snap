// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHEET ## NATIVE ## 01 null 1`] = `null`;

exports[`SHEET ## NATIVE ## 02 {} 1`] = `Object {}`;

exports[`SHEET ## NATIVE ## 03 backgroundColor: 'red' 1`] = `
Object {
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 04 merge 1`] = `null`;

exports[`SHEET ## NATIVE ## 05 merge 1`] = `null`;

exports[`SHEET ## NATIVE ## 06 merge 1`] = `Object {}`;

exports[`SHEET ## NATIVE ## 07 merge 1`] = `
Object {
  "label": Array [
    Array [
      "color: green /*sheet.label*/",
    ],
  ],
  "root": Array [
    Array [
      "backgroundColor: green /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 08 merge 1`] = `
Object {
  "label": Array [
    Array [
      "color: red /*sheet.label*/",
    ],
  ],
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
    Array [
      "backgroundColor: green /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 09 merge 1`] = `
Object {
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
    Array [
      "backgroundColor: blue /*sheet.root/$native*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 10 with theme 1`] = `
Object {
  "label": Array [
    Array [
      "color: white /*sheet.label[0]*/",
      "backgroundColor: darkgreen /*sheet.label[0]*/",
    ],
    Array [
      "margin: 10 /*sheet.label[1]*/",
    ],
  ],
  "root": Array [
    Array [
      "backgroundColor: darkgreen /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 1`] = `
Object {
  "root": Object {
    "$web": Object {
      "color": "green",
    },
    "backgroundColor": "red",
  },
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 2`] = `
Object {
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 3`] = `
Object {
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 4`] = `
Object {
  "root": Array [
    Array [
      "backgroundColor: red /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## NATIVE ## 12 merge ERROR, sheets must be atomized first 1`] = `
Object {
  "label": Object {
    "color": "red",
    "type": "a",
  },
  "root": Array [
    Object {
      "backgroundColor": "red",
    },
    Array [
      "backgroundColor: green /*sheet.root*/",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 01 null 1`] = `null`;

exports[`SHEET ## WEB ## 02 {} 1`] = `Object {}`;

exports[`SHEET ## WEB ## 03 backgroundColor: 'red' 1`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 04 merge 1`] = `null`;

exports[`SHEET ## WEB ## 05 merge 1`] = `null`;

exports[`SHEET ## WEB ## 06 merge 1`] = `Object {}`;

exports[`SHEET ## WEB ## 07 merge 1`] = `
Object {
  "label": Array [
    Array [
      ".b { color:green /*sheet.label*/ }",
    ],
  ],
  "root": Array [
    Array [
      ".a { background-color:green /*sheet.root*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 08 merge 1`] = `
Object {
  "label": Array [
    Array [
      ".b { color:red /*sheet.label*/ }",
    ],
  ],
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
    Array [
      ".c { background-color:green /*sheet.root*/ }",
    ],
  ],
  "webOnly": Array [
    Array [
      ".d { color:green /*sheet.webOnly/$web*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 09 merge 1`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
    Array [
      ".b { color:green /*sheet.root/$web*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 10 with theme 1`] = `
Object {
  "label": Array [
    Array [
      ".d { color:white /*sheet.label[0]*/ }",
      ".c { background-color:darkgreen /*sheet.label[0]*/ }",
    ],
    Array [
      ".e { margin:10px /*sheet.label[1]*/ }",
    ],
  ],
  "root": Array [
    Array [
      ".c { background-color:darkgreen /*sheet.root*/ }",
    ],
  ],
  "webOnly": Array [
    Array [
      ".a { color:darkgray /*theme.secondary.disabled*/ }",
      ".b { background-color:blue /*theme.secondary.disabled*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 1`] = `
Object {
  "root": Object {
    "$web": Object {
      "color": "green",
    },
    "backgroundColor": "red",
  },
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 2`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
    Array [
      ".b { color:green /*sheet.root/$web*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 3`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
    Array [
      ".b { color:green /*sheet.root/$web*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 4`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red /*sheet.root*/ }",
    ],
    Array [
      ".b { color:green /*sheet.root/$web*/ }",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 12 merge ERROR, sheets must be atomized first 1`] = `
Object {
  "label": Object {
    "color": "red",
    "type": "a",
  },
  "root": Array [
    Object {
      "backgroundColor": "red",
    },
    Array [
      ".a { background-color:green /*sheet.root*/ }",
    ],
  ],
  "webOnly": Array [
    Array [
      ".b { color:green /*sheet.webOnly/$web*/ }",
    ],
  ],
}
`;
